### PR TITLE
Better type checking

### PR DESCRIPTION
**Pull request by sbrunner**

### CODE REVIEW

File: geoportal/c2cgeoportal_geoportal/views/printproxy.py [hunk 1]
```diff
@@ -97,32 +104,32 @@ def _capabilities(self, templates, query_string, method):
...
         )
 
     @view_config(route_name="printproxy_report_get")
-    def report_get(self):
+    def report_get(self) -> pyramid.response.Response:
         """ Get the PDF. """
-        url = "{0!s}/report/{1!s}".format(self.config["print_url"], self.request.matchdict.get("ref"))
+        url = Url("{0!s}/report/{1!s}".format(self.config["print_url"], self.request.matchdict.get("ref")))
         if self.config.get("print_get_redirect", False):
             raise HTTPFound(location=url)
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/views/printproxy.py:134]:
```suggestion
            raise HTTPFound(location=url.url)
```

File: geoportal/c2cgeoportal_geoportal/views/theme.py [hunk 2]
```diff
@@ -157,8 +159,11 @@ def _get_metadata(self, item: main.TreeItem, metadata: str, errors: Set[str]):
             )
         )
 
-    def _get_metadatas(self, item: main.TreeItem, errors: Set[str]):
-        metadatas = {}
+    def _get_metadatas(
+        self, item: main.TreeItem, errors: Set[str]
+    ) -> Dict[str, Union[str, int, float, bool, List[Any], Dict[str, Any]]]:
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/views/theme.py:164]: Could help to defined a custom type here as for functionalities.

File: geoportal/c2cgeoportal_geoportal/views/theme.py [hunk 3]
```diff
@@ -1072,14 +1082,20 @@ def get_theme() -> Dict[str, Union[Dict[str, Any], List[str]]]:
...
         @CACHE_REGION.cache_on_arguments()
         def get_theme_anonymous(
-            intranet, interface, sets, min_levels, group, background_layers_group, host
+            intranet: Any,
+            interface: Any,
+            sets: Any,
+            min_levels: Any,
+            group: Any,
+            background_layers_group: Any,
+            host: Any,
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/views/theme.py:1091]: Could be more precise.

**Review by arnaud-morvan — COMMENTED** [review]
See details :

File: docker/qgisserver/geomapfish_qgisserver/accesscontrol.py [hunk 4]
```diff
@@ -426,7 +453,7 @@ def get_area(self, layer, session, read_write=False):
         self.area_cache[key] = (Access.AREA, area)
         return (Access.AREA, area)
 
-    def layerFilterSubsetString(self, layer):  # NOQA
+    def layerFilterSubsetString(self, layer):  # noqa: ignore=N802
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply docker/qgisserver/geomapfish_qgisserver/accesscontrol.py:456]: Missing, thanks :-)

File: geoportal/c2cgeoportal_geoportal/lib/authentication.py [hunk 5]
```diff
@@ -51,12 +53,12 @@
...
 @implementer(IAuthenticationPolicy)
 class UrlAuthenticationPolicy(CallbackAuthenticationPolicy):
-    def __init__(self, aes_key, callback=None, debug=False):
+    def __init__(self, aes_key: str, callback: Callable[[str, Any], List[str]] = None, debug: bool = False):
         self.aeskey = aes_key
         self.callback = callback
         self.debug = debug
 
-    def unauthenticated_userid(self, request):
+    def unauthenticated_userid(self, request: pyramid.request.Request) -> Optional[Dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/lib/authentication.py:61]: :-)

File: geoportal/c2cgeoportal_geoportal/lib/dbreflection.py [hunk 6]
```diff
@@ -55,13 +57,13 @@
...
 
-    def __init__(self, target, value_attr, nullable=True, order_by=None):
+    def __init__(self, target: str, value_attr: str, nullable: bool = True, order_by: str = None):
         self.target = target
         self.value_attr = value_attr
         self.nullable = nullable
         self.order_by = order_by
 
-    def __get__(self, obj, type_=None):
+    def __get__(self, obj: DeclarativeMeta, type_: Any = None) -> Optional[Union["_AssociationProxy", str]]:
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/lib/dbreflection.py:66]: Thanks :-)

File: geoportal/c2cgeoportal_geoportal/views/layers.py [hunk 7]
```diff
@@ -237,21 +239,24 @@ def read_one(self):
...
-    def count(self):
-        set_common_headers(self.request, "layers", NO_CACHE)
+    def count(self) -> int:
+        set_common_headers(self.request, "layers", Cache.no)
 
         protocol = self._get_protocol_for_request()
-        return protocol.count(self.request)
+        count = protocol.count(self.request)
+        if isinstance(count, HTTPException):
+            raise count
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/views/layers.py:248]: For me, it's correct...

**Comment by sbrunner** [comment]
@arnaud-morvan ready for a second pass :-)

**Review by arnaud-morvan — APPROVED** [review]
If you fixed all my previous comments, I just approve.

**Pull request merged**